ops/docker/op-stack-go/Dockerfile

```diff
@@ -23,70 +38,123 @@ COPY . /app
 ARG GIT_COMMIT
 ARG GIT_DATE
 
-ARG CANNON_VERSION=v0.0.0
-ARG OP_PROGRAM_VERSION=v0.0.0
-
-ARG OP_HEARTBEAT_VERSION=v0.0.0
-ARG OP_WHEEL_VERSION=v0.0.0
-
-ARG OP_NODE_VERSION=v0.0.0
-ARG OP_CHALLENGER_VERSION=v0.0.0
-ARG OP_DISPUTE_MON_VERSION=v0.0.0
-ARG OP_BATCHER_VERSION=v0.0.0
-ARG OP_PROPOSER_VERSION=v0.0.0
-ARG OP_CONDUCTOR_VERSION=v0.0.0
-
-
 # separate docker-builds:
 # - op-ufm
 # - proxyd
 # - any JS/TS/smart-contract builds
 
+# TODO: if the build involves optional CGO, then:
+# 1) we may have to disable CGO explicitly
+# 3) apk install cross-build gcc,
+#    and specify to go to use that, with the CC flag
+
 ARG TARGETOS TARGETARCH
 
 # Build the Go services, utilizing caches and share the many common packages.
 # The "id" defaults to the value of "target", the cache will thus be reused during this build.
 # "sharing" defaults to "shared", the cache will thus be available to other concurrent docker builds.
-
+FROM --platform=$BUILDPLATFORM builder as cannon-builder
+ARG CANNON_VERSION=v0.0.0
 RUN --mount=type=cache,target=/root/.cache/go-build cd cannon && make cannon  \
     GOOS=$TARGETOS GOARCH=$TARGETARCH GITCOMMIT=$GIT_COMMIT GITDATE=$GIT_DATE VERSION="$CANNON_VERSION"
+
+FROM --platform=$BUILDPLATFORM builder as op-program-builder
+ARG OP_PROGRAM_VERSION=v0.0.0
```

Comment:
Could we please make sure that we *don't* load the op-program version from git tags? The `op-program` build needs to be exactly reproducible - so if you check out a specific commit and run `make reproducible-prestate` it should produce precisely the same `prestate.json`. If we determine the version from tags, then adding a new tag to the commit would change the result and break dispute games.

This really just applies to the `op-program-client` when built for the MIPS architecture since that's what is used as the prestate but I'd rather leave the version broken (or just always using the commit hash) than risk it being not reproducible.

Reply:
Admittedly `make reproducible-build` doesn't actually go through this docker build process (I think) so it's probably fine...

Reply:
The way the `op-program` [Makefile](https://github.com/ethereum-optimism/optimism/blob/ss/docker-targets-v2/op-program/Makefile#L3) is written, the version cannot be overridden by env vars. It is hardcoded to "v0.0.0":
```
VERSION := v0.0.0
```
I am changing other Makefiles to allow overriding:
```
VERSION ?= v0.0.0
```
...but I will leave the `op-program` Makefile as-is to preserve the hardcoding. Does that work or would you prefer the `op-program-client` to be the sole exception and other binaries built with that same Makefile would allow env var overriding?

Reply:
If it's easy to just have `op-program-client` have a hard coded version and `op-program` has the right version that's probably the ideal case. I'm not really worried about it other than ensuring that the MIPS client build is definitely reproducible.

Reply:
Fixed in https://github.com/ethereum-optimism/optimism/pull/10346/commits/10cf4864633fd6c1a0f1905acdbadc8c343228ab